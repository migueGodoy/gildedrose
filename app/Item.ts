export class Item {  
  name: string;
  sellIn: number;
  quality: number;

  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }

  updateQuality(): void {
    this.quality--;
    this.sellIn--;
  };

}
