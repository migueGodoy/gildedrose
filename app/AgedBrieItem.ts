import { Item } from "./Item";

export class AgedBrieItem extends Item {

    constructor(sellIn: number, quality: number) {
        super('Aged Brie', sellIn, quality);
    }

    updateQuality(): void {
        if (this.quality < 50) {
            this.quality = this.quality + 1
        }
        this.sellIn = this.sellIn - 1;
        if (this.sellIn < 0) {
            if (this.quality < 50) {
                this.quality = this.quality + 1
            }
        }
    }
}