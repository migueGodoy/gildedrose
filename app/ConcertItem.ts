import { Item } from "./Item";

export class ConcertItem extends Item {

    constructor(sellIn: number, quality: number) {
        super('Backstage passes to a TAFKAL80ETC concert', sellIn, quality);
    }

    updateQuality(): void {
        if (this.quality < 50) {
            this.quality = this.quality + 1
            if (this.sellIn < 11) {
                if (this.quality < 50) {
                    this.quality = this.quality + 1
                }
            }
            if (this.sellIn < 6) {
                if (this.quality < 50) {
                    this.quality = this.quality + 1
                }
            }
        }
          
        this.sellIn = this.sellIn - 1;
      
        if (this.sellIn < 0) {
            this.quality = 0
        }
    }
}