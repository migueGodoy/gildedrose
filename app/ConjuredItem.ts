import { Item } from "./Item";

export class ConjuredItem extends Item {
    constructor(name: string, sellIn: number, quality: number) {
        super(name, sellIn, quality);
    }

    updateQuality(): void {
        this.quality -= 2;
        this.sellIn--;
    };
}