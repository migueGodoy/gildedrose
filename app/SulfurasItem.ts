import { Item } from "./Item";

export class SulfurasItem extends Item {

    constructor(sellIn: number) {
        super('Sulfuras, Hand of Ragnaros', sellIn, 80);
    }

    updateQuality(): void {
        
    }
}