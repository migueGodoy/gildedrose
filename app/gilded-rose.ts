import { Item } from "./Item";

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  updateQuality(): Array<Item> {
    for (const currentItem of this.items) {
      currentItem.updateQuality();
    }

    return this.items;
  }
}
