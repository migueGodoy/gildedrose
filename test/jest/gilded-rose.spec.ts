import { AgedBrieItem } from '@/AgedBrieItem';
import { ConcertItem } from '@/ConcertItem';
import { ConjuredItem } from '@/ConjuredItem';
import { GildedRose } from '@/gilded-rose';
import { Item } from "@/Item";
import { SulfurasItem } from '@/SulfurasItem';

describe('Gilded Rose', () => {
  it('should decrease quality by one for normal items', () => {
    const gildedRose = new GildedRose([new Item('foo', 4, 4)]);
    const items = gildedRose.updateQuality();
    expect(items[0].quality).toBe(3);
  });

  it('should increase quality by 2 for AGED BRIE if sellIn falls below 0', () => {
    const gildedRose = new GildedRose([new AgedBrieItem(0, 4)]);
    const items = gildedRose.updateQuality();
    expect(items[0].quality).toBe(6);
  });

  it('should increment quality by 3 for conctert pass with sellIn < 6 and quality < 50', () => {
    const gildedRose = new GildedRose([new ConcertItem(4, 4)]);
    const items = gildedRose.updateQuality();
    expect(items[0].quality).toBe(7);
  });

  it('should increment quality by 2 for conctert pass with sellIn between 6 and 11 and quality < 50', () => {
    const gildedRose = new GildedRose([new ConcertItem(10, 4)]);
    const items = gildedRose.updateQuality();
    expect(items[0].quality).toBe(6);
  });

  it('should not modify Sulfuras quality nor sellIn', () => {
    const gildedRose = new GildedRose([new SulfurasItem(5)]);
    const items = gildedRose.updateQuality();
    expect(items[0].quality).toBe(80);
    expect(items[0].sellIn).toBe(5);
  });

  it('should do a snapshot for 2 days', function () {
    let items = [
      new Item("+5 Dexterity Vest", 10, 20),
      new AgedBrieItem(2, 0),
      new Item("Elixir of the Mongoose", 5, 7),
      new SulfurasItem(0),
      new SulfurasItem(-1),
      new ConcertItem(15, 20),
      new ConcertItem(10, 49),
      new ConcertItem(5, 49),
      new ConjuredItem("Conjured Mana Cake", 3, 6)];

    const gildedRose = new GildedRose(items);

    gildedRose.updateQuality();
    gildedRose.updateQuality();

    expect(items).toMatchSnapshot();
  });

  it('should do a snapshot for 10 days', function () {
    let items = [
      new Item("+5 Dexterity Vest", 10, 20),
      new AgedBrieItem(2, 0),
      new Item("Elixir of the Mongoose", 5, 7),
      new SulfurasItem(0),
      new SulfurasItem(-1),
      new ConcertItem(15, 20),
      new ConcertItem(10, 49),
      new ConcertItem(5, 49),
      new ConjuredItem("Conjured Mana Cake", 3, 6)];

    const gildedRose = new GildedRose(items);

    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();
    gildedRose.updateQuality();

    expect(items).toMatchSnapshot();
  });
});
