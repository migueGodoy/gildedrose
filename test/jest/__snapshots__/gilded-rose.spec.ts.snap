// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gilded Rose should do a snapshot for 2 days 1`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 18,
    "sellIn": 8,
  },
  AgedBrieItem {
    "name": "Aged Brie",
    "quality": 2,
    "sellIn": 0,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": 5,
    "sellIn": 3,
  },
  SulfurasItem {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  SulfurasItem {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  ConcertItem {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 22,
    "sellIn": 13,
  },
  ConcertItem {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 8,
  },
  ConcertItem {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 3,
  },
  ConjuredItem {
    "name": "Conjured Mana Cake",
    "quality": 2,
    "sellIn": 1,
  },
]
`;

exports[`Gilded Rose should do a snapshot for 10 days 1`] = `
Array [
  Item {
    "name": "+5 Dexterity Vest",
    "quality": 10,
    "sellIn": 0,
  },
  AgedBrieItem {
    "name": "Aged Brie",
    "quality": 18,
    "sellIn": -8,
  },
  Item {
    "name": "Elixir of the Mongoose",
    "quality": -3,
    "sellIn": -5,
  },
  SulfurasItem {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": 0,
  },
  SulfurasItem {
    "name": "Sulfuras, Hand of Ragnaros",
    "quality": 80,
    "sellIn": -1,
  },
  ConcertItem {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 35,
    "sellIn": 5,
  },
  ConcertItem {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 50,
    "sellIn": 0,
  },
  ConcertItem {
    "name": "Backstage passes to a TAFKAL80ETC concert",
    "quality": 0,
    "sellIn": -5,
  },
  ConjuredItem {
    "name": "Conjured Mana Cake",
    "quality": -14,
    "sellIn": -7,
  },
]
`;
